# Week 1 Assignment

In [3]:
import numpy as np
import pandas as pd
# importing required libraries

**Load the dataset as a Pandas data frame.**

In [4]:
dataframe = pd.read_csv("Video_Games_Sales_as_at_22_Dec_2016.csv")
# loading csv and saving to pandas dataframe

**Display the first ten rows of data.**

In [5]:
dataframe.head(10)
# viewing first 10 rows of dataframe

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

**Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?**

In [6]:
dataframe.shape
# using shape to see the dimensions of the dataframe 

(16719, 16)

In the context of the data, the shape reveals the dimensions of the dataframe. The returned value (16719, 16) is equivalent to (rows, columns). The columns can be interpreted as different variables while the the rows can be interpreted as observations.

There are 16 columns and 16719 rows in the dataframe


**Find the top five games by critic score.**

In [7]:
dataframe.sort_values(by=["Critic_Score"],ascending=False).head(5)
# acsending by default is set to "True", set to "False" to get the higher critic score to the top

Name Platform  Year_of_Release     Genre  \
227   Tony Hawk's Pro Skater 2       PS           2000.0    Sports   
57         Grand Theft Auto IV      PS3           2008.0    Action   
51         Grand Theft Auto IV     X360           2008.0    Action   
5350               SoulCalibur       DC           1999.0  Fighting   
165         Grand Theft Auto V     XOne           2014.0    Action   

                 Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
227             Activision      3.05      1.41      0.02         0.20   
57    Take-Two Interactive      4.76      3.69      0.44         1.61   
51    Take-Two Interactive      6.76      3.07      0.14         1.03   
5350    Namco Bandai Games      0.00      0.00      0.34         0.00   
165   Take-Two Interactive      2.81      2.19      0.00         0.47   

      Global_Sales  Critic_Score  Critic_Count User_Score  User_Count  \
227           4.68          98.0          19.0        7.7       299.0   
57           10.50          98.0          64.0        7.5      2833.0   
51           11.01          98.0          86.0        7.9      2951.0   
5350          0.34          98.0          24.0        8.8       200.0   
165           5.48          97.0          14.0        7.9       764.0   

                    Developer Rating  
227   Neversoft Entertainment      T  
57             Rockstar North      M  
51             Rockstar North      M  
5350                    Namco      T  
165            Rockstar North      M

**Find the number of video games in the data frame in each genre.**

In [8]:
dataframe['Genre'].unique() 
# reviewing the unique values of 'Genre'

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan], dtype=object)

In [9]:
dataframe['Genre'].value_counts()
# using value_counts() to count the number of games for each genre

Action          3370
Sports          2348
Misc            1750
Role-Playing    1500
Shooter         1323
Adventure       1303
Racing          1249
Platform         888
Simulation       874
Fighting         849
Strategy         683
Puzzle           580
Name: Genre, dtype: int64

My initial thought on this was to use a conditional statement to try and get the answer, luckily I recalled the .value_counts() function. 

**Find the first five games in the data frame on the SNES platform.**

In [10]:
dataframe['Platform'].unique()
# reviewing the unique values of 'Platform' for my own curiosity 

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [11]:
dataframe[(dataframe['Platform'] == "SNES")].head(5)
# filtering the dataframe by platorm 'SNES' and viewing the first 5 rows

Name Platform  Year_of_Release     Genre  \
18                      Super Mario World     SNES           1990.0  Platform   
56                  Super Mario All-Stars     SNES           1993.0  Platform   
71                    Donkey Kong Country     SNES           1994.0  Platform   
76                       Super Mario Kart     SNES           1992.0    Racing   
137  Street Fighter II: The World Warrior     SNES           1992.0  Fighting   

    Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
18   Nintendo     12.78      3.75      3.54         0.55         20.61   
56   Nintendo      5.99      2.15      2.12         0.29         10.55   
71   Nintendo      4.36      1.71      3.00         0.23          9.30   
76   Nintendo      3.54      1.24      3.81         0.18          8.76   
137    Capcom      2.47      0.83      2.87         0.12          6.30   

     Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
18            NaN           NaN        NaN         NaN       NaN    NaN  
56            NaN           NaN        NaN         NaN       NaN    NaN  
71            NaN           NaN        NaN         NaN       NaN    NaN  
76            NaN           NaN        NaN         NaN       NaN    NaN  
137           NaN           NaN        NaN         NaN       NaN    NaN

**Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.**

In [12]:
dataframe['Publisher'].unique()
# reviewing the unique values of 'Publisher' for my own curiosity

array(['Nintendo', 'Microsoft Game Studios', 'Take-Two Interactive',
       'Sony Computer Entertainment', 'Activision', 'Ubisoft',
       'Bethesda Softworks', 'Electronic Arts', 'Sega', 'SquareSoft',
       'Atari', '505 Games', 'Capcom', 'GT Interactive',
       'Konami Digital Entertainment', 'Square Enix',
       'Sony Computer Entertainment Europe', 'Virgin Interactive',
       'LucasArts', 'Warner Bros. Interactive Entertainment',
       'Universal Interactive', 'Eidos Interactive', 'RedOctane',
       'Vivendi Games', 'Enix Corporation', 'Namco Bandai Games',
       'Palcom', 'Hasbro Interactive', 'THQ', 'Fox Interactive',
       'Acclaim Entertainment', 'MTV Games', 'Disney Interactive Studios',
       nan, 'Codemasters', 'Majesco Entertainment', 'Red Orb', 'Level 5',
       'Arena Entertainment', 'Midway Games', 'JVC', 'Deep Silver',
       'NCSoft', '989 Studios', 'UEP Systems', 'Parker Bros.', 'Maxis',
       'Imagic', 'Tecmo Koei', 'ASCII Entertainment', 'Valve Software',


In [25]:
dataframe.groupby(['Publisher'], sort=False)['Global_Sales'].sum().head(15)
# grouping by 'Publisher', sort=False sorts the values in descedning order of the total sum()

Publisher
Nintendo                        1788.81
Microsoft Game Studios           248.32
Take-Two Interactive             403.82
Sony Computer Entertainment      606.48
Activision                       731.16
Ubisoft                          471.61
Bethesda Softworks                85.05
Electronic Arts                 1116.96
Sega                             270.35
SquareSoft                        57.65
Atari                            156.83
505 Games                         55.32
Capcom                           200.02
GT Interactive                    25.23
Konami Digital Entertainment     282.39
Name: Global_Sales, dtype: float64

I had no idea there would be so many publishers of games, it would be interesting to see how the number of publishers has changed over time.

**Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.**

In [14]:
# dataframe['Percent_NA_Sales'] = dataframe['NA_Sales']/dataframe['Global_Sales']
# dataframe.head(5)

I decided to leave the existing dataframe as is and create a new dataframe (dataframe2) with the additional column.

In [15]:
x = dataframe['NA_Sales']/dataframe['Global_Sales']
# setting results to a variable 
dataframe2 = dataframe.assign(Percent_NA_Sales = x)
# assigning the variable a name and adding the column to the new data frame
dataframe2.head(5)
# viewing the first five rows of the data frame 

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  Percent_NA_Sales  
0          51.0          8       322.0  Nintendo      E          0.501151  
1           NaN        NaN         NaN       NaN    NaN          0.722664  
2          73.0        8.3       709.0  Nintendo      E          0.441441  
3          73.0          8       192.0  Nintendo      E          0.476350  
4           NaN        NaN         NaN       NaN    NaN          0.359260

In [16]:
dataframe.head(5)
# verifiying the original dataframe does not have the new column 

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  
0          51.0          8       322.0  Nintendo      E  
1           NaN        NaN         NaN       NaN    NaN  
2          73.0        8.3       709.0  Nintendo      E  
3          73.0          8       192.0  Nintendo      E  
4           NaN        NaN         NaN       NaN    NaN

**Find the number NaN entries (missing data values) in each column.**

In [17]:
dataframe.isnull().sum(axis = 0)
# using isnull() to find the NaN values and .sum() to sum the total NaN by column

Name                  2
Platform              0
Year_of_Release     269
Genre                 2
Publisher            54
NA_Sales              0
EU_Sales              0
JP_Sales              0
Other_Sales           0
Global_Sales          0
Critic_Score       8582
Critic_Count       8582
User_Score         6704
User_Count         9129
Developer          6623
Rating             6769
dtype: int64

**Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.**

In [18]:
y = dataframe['User_Score'].replace('tbd', np.nan)
# saving the returned pandas series as a variable 

In [19]:
y.unique()
# verifying that the string 'tbd' no longer is present

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', '8.8', '6.9', '9.4', '6.8', '6.1',
       '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6', '3.7',
       '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9', '3.1',
       '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3', '2.7',
       '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0', '1.6',
       '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2', '2.3',
       '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'], dtype=object)

In [20]:
dataframe = dataframe.assign(User_Score = y)
# assigning the pandas series 'y' as a new column (overwriting the existing) 'User_Score'
dataframe['User_Score'].unique()
# checking the unique values of 'User_Score' to be sure the string 'tbd' still does not exist

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', '8.8', '6.9', '9.4', '6.8', '6.1',
       '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6', '3.7',
       '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9', '3.1',
       '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3', '2.7',
       '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0', '1.6',
       '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2', '2.3',
       '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'], dtype=object)

In [21]:
dataframe['User_Score'].median()
# using median to calculate the median 'User_Score'

7.5

In [26]:
z = dataframe['User_Score'].replace(np.nan, '7.5')
# using replace() to repalce NaN values with 7.5 which was the calcualted median and saving to variable 'z'
dataframe = dataframe.assign(User_Score = z)
# assignign the new pandas series saved to z as 'User_Score'
dataframe.head(20)
# Viewing the first 20 rows to check the column 'User_Score'

Name Platform  Year_of_Release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           Genre               Publisher  NA_Sales  EU_Sales  JP_Sales  \
0         Sports                Nintendo     41.36     28.96      3.77   
1       Platform                Nintendo     29.08      3.58      6.81   
2         Racing                Nintendo     15.68     12.76      3.79   
3         Sports                Nintendo     15.61     10.93      3.28   
4   Role-Playing                Nintendo     11.27      8.89     10.22   
5         Puzzle                Nintendo     23.20      2.26      4.22   
6       Platform                Nintendo     11.28      9.14      6.50   
7           Misc                Nintendo     13.96      9.18      2.93   
8       Platform                Nintendo     14.44      6.94      4.70   
9        Shooter                Nintendo     26.93      0.63      0.28   
10    Simulation                Nintendo      9.05     10.95      1.93   
11        Racing                Nintendo      9.71      7.47      4.13   
12  Role-Playing                Nintendo      9.00      6.18      7.20   
13        Sports                Nintendo      8.92      8.03      3.60   
14          Misc  Microsoft Game Studios     15.00      4.89      0.24   
15        Sports                Nintendo      9.01      8.49      2.53   
16        Action    Take-Two Interactive      7.02      9.09      0.98   
17        Action    Take-Two Interactive      9.43      0.40      0.41   
18      Platform                Nintendo     12.78      3.75      3.54   
19          Misc                Nintendo      4.74      9.20      4.16   

    Other_Sales  Global_Sales  Critic_Score  Critic_Count User_Score  \
0          8.45         82.53          76.0          51.0          8   
1          0.77         40.24           NaN           NaN        7.5   
2          3.29         35.52          82.0          73.0        8.3   
3          2.95         32.77          80.0          73.0          8   
4          1.00         31.37           NaN           NaN        7.5   
5          0.58         30.26           NaN           NaN        7.5   
6          2.88         29.80          89.0          65.0        8.5   
7          2.84         28.92          58.0          41.0        6.6   
8          2.24         28.32          87.0          80.0        8.4   
9          0.47         28.31           NaN           NaN        7.5   
10         2.74         24.67           NaN           NaN 

I feel that I completed the last task correctly for this assignment, but I am unsure if it was the best way to do it. When I was attempting this task in other ways I noticed the only way I could get the values to be replaced was to save the results to a variable. 